dlib/matrix/kiss_fft.h

```diff
@@ -0,0 +1,733 @@
+        {
+            int dimReal  = plan.cfg_r.substate.nfft*2; //recall the real fft size is half the length of the input
+            int dimOther = plan.cfg_nd.dimprod();
+            int nrbins   = dimReal/2+1;
+
+            std::vector<std::complex<T>> tmp1(std::max<int>(nrbins, dimOther));
+            std::vector<std::complex<T>> tmp2(plan.cfg_nd.dimprod()*dimReal);
+
+            // take a real chunk of data, fft it and place the output at correct intervals
+            for (int k1 = 0; k1 < dimOther; ++k1) 
+            {
+                kiss_fftr(plan.cfg_r, timedata + k1*dimReal , &tmp1[0]); // tmp1 now holds nrbins complex points
+                for (int k2 = 0; k2 < nrbins; ++k2)
+                   tmp2[k2*dimOther+k1] = tmp1[k2];
+            }
+
+            for (int k2 = 0; k2 < nrbins; ++k2) 
+            {
+                kiss_fftnd(plan.cfg_nd, &tmp2[k2*dimOther], &tmp1[0]);  // tmp1 now holds dimOther complex points
+                for (int k1 = 0; k1 < dimOther; ++k1) 
+                    freqdata[ k1*(nrbins) + k2] = tmp1[k1];
+            }
+        }
+
+        template<typename T>
+        void kiss_fftndri(const kiss_fftndr_state<T>& plan, const std::complex<T>* freqdata, T* timedata)
+        {
+            int dimReal  = plan.cfg_r.substate.nfft*2; //recall the real fft size is half the length of the input
+            int dimOther = plan.cfg_nd.dimprod();
+            int nrbins   = dimReal/2+1;
+
+            std::vector<std::complex<T>> tmp1(std::max<int>(nrbins, dimOther));
+            std::vector<std::complex<T>> tmp2(plan.cfg_nd.dimprod()*dimReal);
+
+            for (int k2 = 0; k2 < nrbins; ++k2) 
+            {
+                for (int k1 = 0; k1 < dimOther; ++k1) 
+                    tmp1[k1] = freqdata[ k1*(nrbins) + k2 ];
+                kiss_fftnd(plan.cfg_nd, &tmp1[0], &tmp2[k2*dimOther]);
+            }
+
+            for (int k1 = 0; k1 < dimOther; ++k1) 
+            {
+                for (int k2 = 0; k2 < nrbins; ++k2)
+                    tmp1[k2] = tmp2[ k2*dimOther+k1 ];
+                kiss_fftri(plan.cfg_r, &tmp1[0], timedata + k1*dimReal);
+            }
+        }
+
+        struct hasher
+        {
+            size_t operator()(const plan_key& key) const {return key.hash();}
+        };
+
+        template<typename plan_type>
+        class cache
+        {
+        public:
+            static cache& get()
+            {
+                static cache singleton;
+                return singleton;
+            }
+
+            void get_plan(const plan_key& key, plan_type& plan)
+            {
+                std::lock_guard<std::mutex> l(m);
+                auto it = plans.find(key);
+                if (it != plans.end())
+                {
+                    plan = it->second;
+                }
+                else
+                {
+                    plans[key] = plan_type(key);
+                    plan = plans[key];
```

Comment:
This is making a copy.  Why not return a reference to this?

Reply:
If i return a reference, then another thread comes in and creates a new plan, then the reference to the first object may potentially be incorrect no? Or do unordered_maps not reallocate memory when there is an insertion?

Reply:
According to cppreference : ```References and pointers to either key or data stored in the container are only invalidated by erasing that element, even when the corresponding iterator is invalidated.```. So it's ok to return references?

Reply:
Right, so long as you don't erase the element it's fine to return a reference.  And this copy is expensive.  It's copying a bunch of stuff and doing memory allocations and all that.